### ISSUE TITLE
XPUB-XSUB high water mark reached no matter the value of the HWM

### ISSUE THREAD

**Issue by BogdanEM** [zeromq/jeromq#238]
I'm trying to understand a bit how the HWM settings need to be configured for a Pub-Sub scenario so that no messages are dropped. 

Right now I have:
- a XPUB socket with a 60000 SndHWM connected to...
- a XSUB socket with a 60000 RcvHWM

and **I'm losing messages**. 

Messages sent by the publisher (less than 1k in size most of them) are marked with a serial number. The subscriber checks the serial number to detect gaps. It is absolutely mandatory for my use case that no messages are lost.

Publisher and subscriber run at similar speeds (subscriber is actually faster) with spikes of around 30000 messages/second but most of the time slower, with a few thousand messages/second.

I believed that a 60000 size HWM for sender and subscriber should cover the spikes but my subscriber detects gaps. I even tried 10 million on each socket HWM and that too loses messages. As I said, subscriber is fast and consumes faster than what the publisher can produce.

I tried to add some logs to see what's happening and it's the XPUB that reaches the HWM, no matter the value of the HWM. 

I added some logs in `Pipe.java`, in `check_write` to write the values of this test:

```
boolean full = hwm > 0 && msgs_written - peers_msgs_read == (long) (hwm);
```

and also in `ZObject.java` in `process_command` on `ACTIVATE_WRITE` to see what's going on with the `cmd_.arg` parameter and this is what the logs show:

```
diff: 0, msgs_written: 0, peers_msgs_read: 0
diff: 1, msgs_written: 1, peers_msgs_read: 0
diff: 2, msgs_written: 2, peers_msgs_read: 0
diff: 3, msgs_written: 3, peers_msgs_read: 0
diff: 4, msgs_written: 4, peers_msgs_read: 0
...
...
diff: 59293, msgs_written: 59293, peers_msgs_read: 0
diff: 59294, msgs_written: 59294, peers_msgs_read: 0
diff: 59295, msgs_written: 59295, peers_msgs_read: 0
diff: 59296, msgs_written: 59296, peers_msgs_read: 0
process_activate_write called with 58976
diff: 321, msgs_written: 59297, peers_msgs_read: 58976
diff: 322, msgs_written: 59298, peers_msgs_read: 58976
diff: 323, msgs_written: 59299, peers_msgs_read: 58976
diff: 324, msgs_written: 59300, peers_msgs_read: 58976
...
...
diff: 58979, msgs_written: 117955, peers_msgs_read: 58976
diff: 58980, msgs_written: 117956, peers_msgs_read: 58976
diff: 58981, msgs_written: 117957, peers_msgs_read: 58976
diff: 58982, msgs_written: 117958, peers_msgs_read: 58976
diff: 58983, msgs_written: 117959, peers_msgs_read: 58976
diff: 58984, msgs_written: 117960, peers_msgs_read: 58976
diff: 58985, msgs_written: 117961, peers_msgs_read: 58976
diff: 58986, msgs_written: 117962, peers_msgs_read: 58976
diff: 58987, msgs_written: 117963, peers_msgs_read: 58976
process_activate_write called with 117952
...
...
```

I would have expected the `diff` to remain low most of the times but it just increases until it's close to the HWM then decreases, then it increases and then it decreases, again and again until at some point it goes up to the HWM, the test above shows "full" and my subscriber detects gaps. I also don't understand why the `process_activate_write` is called so rarely.

So my questions would be:
- how do I set up HWM values for both sockets, for the case described  above;
- what's the meaning of the `58976` value for `process_activate_write`'s `cmd_.arg`s parameter;
- is the log output normal? As I said, I would have expected the `diff` to remain at lower values since messages are read faster than they are written;
- what am I doing wrong :(?

I'm using Jeromq 0.3.4 on Windows7.

Thank you!

**Comment by BogdanEM** [comment]
Debugged a bit the code to get a feeling of what's happening and this is my understanding so far:

1. there is a XPUB pipe sending messages to a XPUBSession pipe;
2. from time to time according to `Pipe#read()`, the XPUBSession pipe acknowledges back (ACTIVATEWRITE message) to the XPUB pipe the number of messages it has received so far;

    `if (lwm > 0 && msgs_read % lwm == 0) { `
    `   send_activate_write(peer, msgs_read); `
    `}`

3. XPUB pipe is owned by the main thread, XPUBSession pipe is owned by the IO thread;
4. the code from point 2) uses the LWM which is computed with the `Pipe#compute_lwm` based on the HWM. For example, a HWM=60000 results in LWM=58976, basically HWM - 1024.

Now, since the ACTIVATEWRITE message is sent between different parallel executing threads can the following scenario happen:

1. XPUB pipe sends messages 1 to 58976 to the XPUBSession session;
2. XPUBSession upon receiving the 58976th message sends back an ACTIVATEWRITE to the XPUB pipe;
3. before the XPUB pipe has a chance to process the ACTIVATEWRITE command, a spike occurs and the XPUB pipe sends 1024 messages. At this point the test for "full" becomes true and messages are dropped when in fact they shouldn't be.

Is this scenario correct? Should the LWM be involved when sending the ACTIVATEWRITE command in the `Pipe#read()` code?

**Comment by BogdanEM** [comment]
I eventually went with infinite HWM and things are running without issues. Unless I'm missing something, I can only conclude that the mechanism I described above is broken since it relies on lucky timing to work.

**Comment by fredoboulo** [comment]
I found that issue recently: the computation of LWM was responsible for that effect, specifically in Pipe#computeLwm, when the hwm is bigger than 1024 (I did not quantify the exact threshold when that happened). As it looks that you are able to modify the sources, you could try to modify this method by returning (hwm+1)/2 every time.

The source code in master has that modification already, so it should be available for next release.

**Comment by daveyarwood** [comment]
Based on the release dates, I believe this is fixed as of 0.4.1.
